src/CreateSignalFactory.js

```diff
@@ -150,64 +137,71 @@ module.exports = function (controller, model, services, compute, modules) {
               runBranch(branch, index + 1)
             } else {
               var promises = currentBranch.map(function (action) {
-                controller.emit('actionStart', {action: action, signal: signal})
+                var resolver = null
+                var promise = new Promise(function (resolve) {
+                  resolver = resolve
+                })
+                controller.emit('actionStart', {
+                  action: action,
+                  signal: signal,
+                  options: options,
+                  payload: payload
+                })
                 var actionFunc = actions[action.actionIndex]
-                var inputArg = actionFunc.defaultInput ? utils.merge({}, actionFunc.defaultInput, signalArgs) : signalArgs
-                var actionArgs = createActionArgs.async(action, inputArg, model, compute, services, Object.keys(modules))
 
-                if (utils.isDeveloping() && actionFunc.input) {
-                  utils.verifyInput(action.name, signal.name, actionFunc.input, inputArg)
-                }
+                var contextProviders = [
+                  inputProvider,
+                  stateProvider,
+                  servicesProvider,
+                  outputProvider
+                ].concat(utils.extractExternalContextProviders(externalContextProviders, options.modulePath))
+                .concat(options.context || [])
```

Comment:
do we still want to allow define context providers for each signal?
I thought we decided to keep options clean from "special options". Also it introduces extra way to register context provider which is redundant.